WordPress/gutenberg
# Backmerge block renaming fixes/refactors from 6.4 branch into Gutenberg trunk

## What?



Back merges the refactors that were undertaken directly in the `wp/6.4` branch to improve Block Renaming support.

Namely:

- https://github.com/WordPress/gutenberg/pull/55604
- Closes https://github.com/WordPress/gutenberg/issues/55338

Also fixes [this Issue report](https://github.com/WordPress/gutenberg/issues/55643#issuecomment-1812739942).

Part of https://github.com/WordPress/gutenberg/issues/55643

## Why?


The WP 6.4 branch got ahead of this branch due to the nature of some bugfixes that had to land _directly_ in the 6.4 branch because Gutenerg trunk had already progressed during the 6.4 Beta/RC cycle.

So now we bring Gutenberg back into line with the improvements in 6.4 whilst also retaining the improvements already in `trunk`.

## How?


- Cherry pick changes from 6.4 branch.
- Fix merge conflicts and ensure functionality already in Gutenberg `trunk` is restored.
- Improve e2e tests - I have tried to strike a balance between "coverage" and avoiding creating lots of e2e tests (which are inherently "expensive"). If there are any scenarios you think I've missed please say.

## Testing Instructions





- New Post
- Add blocks:
    - Group
    - Paragraph
    - Heading
    - Quote
- Toggle list view
- Click blocks options menu for each block in list view and check that Rename works as expected (can rename, remove the custom name...etc).
- Toggle Inspector Controls for each blocks and check that the rename control udner `Advanced` panel also behaves as expected and changes are synchronised to the List View.
- Now check the Rename option appears even when the block is not selected:
    - select Paragraph
    - click the block options menu for the `Heading` block (**note:** the block that is _not_ selected).
    - verify Rename menu item is shown and works as expected.
- Now verify Query Block behaviour
    - go to Post Template inside Query block add wrap contents in a Group block
    - click the block options menu for that Group and check you only see the `Rename` option _once_ in the menu.
- Now check the following blocks do _not_ show the rename control on either List View menu _or_ in Inspector Controls:
    - core/block
    - core/template-part
    - core/pattern
    - core/navigation

### Testing Instructions for Keyboard


## Screenshots or screencast
